### PR TITLE
[WIP] External Aeron Sink/Source

### PR DESCRIPTION
**Pull request by chbatey**
Hacked on this during travel. Pulled out most of the duplication with Artery, can't be the same impl as public is typed to ByteString and internal is a pooled buffer. 

WIll finish off if we think it is useful.

Refs https://github.com/akka/akka/issues/26281

### CODE REVIEW

File: akka-aeron/src/main/resources/reference.conf [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    # will be permanently failed and any streams created will be failed
+    driver-timeout = 10s
+    # If set to "<embedded>" an emebedded media driver will be started
+    # during connection creation. Set to the aeron directory location if
+    # using an external media driver
+    aeron-directory = "<embedded>"
+
+  }
+
+  # Configuration for running an embedded media driver
```
Comment (chbatey) [inline akka-aeron/src/main/resources/reference.conf]: TODO

File: akka-actor/src/main/scala-2.12/akka/util/ByteIterator.scala [hunk 2]
```diff
@@ -264,8 +264,8 @@ object ByteIterator {
 
     final override def takeWhile(p: Byte ⇒ Boolean): this.type = {
       var stop = false
-      var builder = new ListBuffer[ByteArrayIterator]
-      while (!stop && !iterators.isEmpty) {
+      val builder = new ListBuffer[ByteArrayIterator]
+      while (!stop && iterators.nonEmpty) {
```
Comment (helena) [inline akka-actor/src/main/scala-2.12/akka/util/ByteIterator.scala]: 👍

File: akka-aeron/src/main/resources/reference.conf [hunk 3]
```diff
@@ -0,0 +1,23 @@
+akka.aeron {
+  # Connnection settings for Aeron
+  connection {
+    # Timeout for communication with the local media driver. If exceeded the connection
+    # will be permanently failed and any streams created will be failed
+    driver-timeout = 10s
+    # If set to "<embedded>" an emebedded media driver will be started
+    # during connection creation. Set to the aeron directory location if
+    # using an external media driver
+    aeron-directory = "<embedded>"
```
Comment (helena) [inline akka-aeron/src/main/resources/reference.conf]: Should this just be 
```suggestion
    aeron-directory = "embedded"
```

File: akka-aeron/src/main/resources/reference.conf [hunk 4]
```diff
@@ -0,0 +1,23 @@
+akka.aeron {
+  # Connnection settings for Aeron
+  connection {
+    # Timeout for communication with the local media driver. If exceeded the connection
+    # will be permanently failed and any streams created will be failed
```
Comment (helena) [inline akka-aeron/src/main/resources/reference.conf]: Doc the failure behavior?

File: akka-aeron/src/main/resources/reference.conf [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+  }
+
+  # Configuration for running an embedded media driver
+  media-driver {
+
+  }
+
+  # Low level settings for tuning Aeron threading model
+  advanced {
```
Comment (helena) [inline akka-aeron/src/main/resources/reference.conf]: What's the rationale behind separating these as opposed to say this, and are these not applicable to inside the `media-driver`? Just checking.
```suggestion
  threading {
    some-logical = defaults
```

File: akka-aeron/src/main/scala/akka/aeron/AeronExtension.scala [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object AeronSettings {
+
+  def apply(config: Config): AeronSettings = {
+    new AeronSettings(
+      config.getDuration("connection.driver-timeout").toMillis,
+      config.getString("connection.aeron-directory") match {
```
Comment (helena) [inline akka-aeron/src/main/scala/akka/aeron/AeronExtension.scala]: Not a personal fan of config work outside a settings class, mixed in with the regular code

Comment (chbatey) [reply akka-aeron/src/main/scala/akka/aeron/AeronExtension.scala]: I usually like to have the settings class not be aware of a `Config` so they are easy to construct for tests and then have the reading from a `Config` in an apply. Do you see any benefits of putting the config work inside the settings class?

**Review by patriknw — COMMENTED** [review]
I think this is looking good and something we could do as part of Akka 2.6. It might be somewhat unclear when we would recommend using these since there is some use case overlap with StreamRefs. Anyway we can have this akka-aeron module. Good to break out some of the Aeron specific things for akka-remote anyway. Then it could possibly evolve with more things as we have talked about.

**Comment by chbatey** [comment]
:+1: let's keep this on hold until 2.6

**Comment by chbatey** [comment]
This isn't ready for review yet but I updated it with master

**Comment by johanandren** [comment]
Let's close this for now and potentially pick it up in the future.

**Pull request closed without merging**